src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp

```diff
@@ -2848,211 +2925,470 @@ __ vpternlogq(dst1, 0x96, src12, src13, Assembler::AVX_512bit); \
   }
+  carrylessMultiply(ZTMP4, ZTMP5, ZTMP6, ZTMP7, ZTMP9, ZTMPA, ZTMP8);
+
+  //update sums
+  xorGHASH(GL, GH, GL, GH, ZTMP0, ZTMP2, ZTMP1, ZTMP3, ZTMP6, ZTMP4, ZTMP7, ZTMP5);
+
+  if (do_reduction) {
+  //new reduction
+    __ evmovdquq(xmm23, ExternalAddress(ghash_polynomial_addr()), Assembler::AVX_512bit, rbx /*rscratch*/);
+    __ evpclmulqdq(HASH, GL, xmm23, 0x10, Assembler::AVX_512bit);
+    __ vpshufd(ZTMP0, GL, 78, Assembler::AVX_512bit);
+    __ vpternlogq(HASH, 0x96, GH, ZTMP0, Assembler::AVX_512bit);
+    if (do_hxor) {
+      vhpxori4x128(HASH, ZTMP0);
+    }
+  }
+}
+
+//Stitched GHASH of 16 blocks(with reduction) with encryption of N blocks
+//followed with GHASH of the N blocks.
+void StubGenerator::gcm_enc_dec_last_avx512(Register len, Register in, Register pos, XMMRegister HASH, Register subkeyHtbl,
+                                            int ghashin_offset, int hashkey_offset, bool start_ghash, bool do_reduction) {
+  //there is 0 blocks to cipher so there are only 16 blocks for ghash and reduction
+  ghash16_avx512(start_ghash, do_reduction, false, false, true, in, pos, subkeyHtbl, HASH, ghashin_offset, 0, 0, hashkey_offset);
+  //**ZT01 may include sensitive data
+}
+
+//Main GCM macro stitching cipher with GHASH
+//encrypts 16 blocks at a time
+//ghash the 16 previously encrypted ciphertext blocks
+void StubGenerator::ghash16_encrypt_parallel16_avx512(Register in, Register out, Register ct, Register pos, Register avx512_subkeyHtbl,
+                                                      Register CTR_CHECK, Register NROUNDS, Register key, bool hk_broadcast,
+                                                      bool is_hash_start, bool do_hash_reduction, bool do_hash_hxor, bool no_ghash_in,
+                                                      int ghashin_offset, int aesout_offset, int hashkey_offset) {
+  const XMMRegister B00_03 = xmm0;
+  const XMMRegister CTR_BE = xmm2;
+  const XMMRegister B04_07 = xmm3;
+  const XMMRegister B08_11 = xmm4;
+  const XMMRegister B12_15 = xmm5;
+  const XMMRegister SHFMSK = xmm29;
+  const XMMRegister THH1 = xmm6;
+  const XMMRegister THL1 = xmm7;
+  const XMMRegister TLH1 = xmm10;
+  const XMMRegister TLL1 = xmm11, THH2 = xmm12, THL2 = xmm13, TLH2 = xmm15;
+  const XMMRegister TLL2 = xmm16, THH3 = xmm17, THL3 = xmm19, TLH3 = xmm20;
+  const XMMRegister TLL3 = xmm21, DATA1 = xmm17, DATA2 = xmm19, DATA3 = xmm20, DATA4 = xmm21;
+  const XMMRegister AESKEY1 = xmm30, AESKEY2 = xmm31;
+  const XMMRegister GHKEY1 = xmm1, GHKEY2 = xmm18, GHDAT1 = xmm8, GHDAT2 = xmm22;
+  const XMMRegister ADDBE_4x4 = xmm27, ADDBE_1234 = xmm28;
+  const XMMRegister GHASH_IN = xmm14, TO_REDUCE_L = xmm25, TO_REDUCE_H = xmm24;
+  const int hkey_gap = 16 * 32;
+
+  Label blocks_overflow, blocks_ok, skip_shuffle, cont, aes_256, aes_192, last_aes_rnd;
+
+  __ cmpl(CTR_CHECK, (256 - 16));
+  __ jcc(Assembler::greaterEqual, blocks_overflow);
```

Comment:
Should this be:
__ cmpb(CTR_CHECK, (256 - 16)
__ jcc(Assembler::aboveEqual, blocks_overflow);